Gems/ScriptCanvas/Code/Include/ScriptCanvas/AutoGen/ScriptCanvasAutoGenRegistry.h

```diff
@@ -6,89 +6,168 @@
+class IScriptCanvasNodeFactory
+{
+public:
 
-    class ScriptCanvasRegistry
+    virtual const AZStd::string GetGemOrModuleName() const = 0;
+    virtual const AZStd::string GetName() const = 0;
+    virtual const AZStd::string GetID() const = 0;
+
+    virtual AZ::ComponentDescriptor* GetDescriptor() const = 0;
+};
+
+//! Holds the complete list of Script Canvas nodes (grammar and nodeables), all
+//! gems and modules will register their nodes into this model
+class ScriptCanvasModel
+{
+public:
+
+    static ScriptCanvasModel& Instance();
+
+    ScriptCanvasModel() = default;
+    virtual ~ScriptCanvasModel()
     {
-    public:
-        virtual ~ScriptCanvasRegistry() = default;
-        virtual void Init([[maybe_unused]] NodeRegistry* nodeRegistry) {}
-        virtual void Reflect([[maybe_unused]] AZ::ReflectContext* context) {}
-        virtual AZStd::vector<AZ::ComponentDescriptor*> GetComponentDescriptors() { return {}; }
-        void ReleaseDescriptors();
-    protected:
-        AZStd::vector<AZ::ComponentDescriptor*> m_cachedDescriptors;
+        m_registeredReflections.clear();
+        m_registry.clear();
+    }
+
+    void Init();
+    void Register(const char* gemOrModuleName, const char* typeName, const char* typeHash, IScriptCanvasNodeFactory* factory);
+
+    using ReflectFunction = AZStd::function<void(AZ::ReflectContext*)>;
+    void RegisterReflection(const AZStd::string& name, ReflectFunction reflect);
+
+    void Reflect(AZ::ReflectContext* context);
+
+    const AZ::ComponentDescriptor* GetDescriptor(const char* typehash) const;
+
+private:
+
+    // Minimum data required for runtime operations
+    struct Entry
+    {
+        AZStd::string m_gemOrModuleName;
+        IScriptCanvasNodeFactory* m_factory;
+        AZ::ComponentDescriptor* m_descriptor;
     };
 
-    //! AutoGenRegistryManager
-    //! The registry manager contains all autogen functions, nodeables and grammars metadata which will be registered
-    //! for ScriptCanvas
-    class AutoGenRegistryManager
+    using RegisteredItems = AZStd::unordered_map<AZStd::string, Entry>;
+    RegisteredItems m_registry;
+
+    AZStd::unordered_map<AZStd::string, ReflectFunction> m_registeredReflections;
+
+    bool m_verbose = false;
+
+public:
+
+    const RegisteredItems& GetEntries() const
     {
-    public:
-        AutoGenRegistryManager() = default;
-        ~AutoGenRegistryManager();
+        return m_registry;
+    }
 
-        static AutoGenRegistryManager* GetInstance();
+};
 
-        //! Init all AutoGen registries
-        static void Init();
+//! Helper object to compute the hash of a given string
+class ScriptCanvasNodeHash
+{
+public:
 
-        //! Init specified AutoGen registry by given name
-        static void Init(const char* registryName);
+    static inline AZStd::string Encode(AZStd::string input)
+    {
+        AZStd::hash<AZStd::string> hash_string;
+        auto h = hash_string(input);
+        return AZStd::string::format("%lX", h);
```

Comment:
This is not the correct formatter.
`%lx` is only applicable on Linux/Android where size_t has a type of `unsigned long`.
It is not applicable on MacOS, iOS and Windows wheer size_t has a type of `unsigned long long`.

The correct formatter here is `%zX`
```suggestion
        return AZStd::string::format("%zX", h);
```

Reply:
thank you!